Identify/decide minimum supported version of Rust

C.f. #1238 

If I understand correctly:

https://github.com/rust-lang/git2-rs/blob/7415113423d3b852e3527a62a97a8505ad9a7fe5/Cargo.toml#L16 means that MSRV >= 1.56.0

https://github.com/rust-lang/git2-rs/blob/7415113423d3b852e3527a62a97a8505ad9a7fe5/Cargo.toml#L20 means MSRV >= 1.56.0 (https://docs.rs/crate/bitflags/2.1.0/source/Cargo.toml#55)

https://github.com/rust-lang/git2-rs/blob/7415113423d3b852e3527a62a97a8505ad9a7fe5/Cargo.lock#L474-L476 means MSRV >= 1.63.0 (https://docs.rs/crate/libc/0.2.174/source/Cargo.toml#52)

https://github.com/rust-lang/git2-rs/blob/7415113423d3b852e3527a62a97a8505ad9a7fe5/src/cred.rs#L301 means MSRV >= 1.70.0 (https://blog.rust-lang.org/2023/06/01/Rust-1.70.0/) (unless we want to say that there is a different MSRV depending on which features are enabled?)

https://github.com/rust-lang/git2-rs/blob/7415113423d3b852e3527a62a97a8505ad9a7fe5/src/opts.rs#L527 means *tests* are MSRV >= 1.70.0 for the same reason

* cargo lockfile version 4 is default since 1.83, supported since 1.78, https://doc.rust-lang.org/nightly/cargo/CHANGELOG.html#cargo-183-2024-11-28
* str::from_utf8() is 1.87.0+, https://blog.rust-lang.org/2025/05/15/Rust-1.87.0/
 
```
root@406fa489d8c6:/rust/git2-rs# cargo install cargo-msrv --locked
...
root@406fa489d8c6:/rust/git2-rs# cargo msrv find
  [Meta]   cargo-msrv 0.19.3  

Compatibility Check #1: Rust 1.76.0
  [FAIL]   Is incompatible 
                                                                                                                                         
  ╭─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
  │ error: failed to parse lock file at: /rust/git2-rs/Cargo.lock                                                                       │
  │                                                                                                                                     │
  │ Caused by:                                                                                                                          │
  │   lock file version `4` was found, but this version of Cargo does not understand this lock file, perhaps Cargo needs to be updated? │
  │                                                                                                                                     │
  ╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
                                                                                                                                         

Compatibility Check #2: Rust 1.86.0
  [FAIL]   Is incompatible 
                                                                                                          
  ╭──────────────────────────────────────────────────────────────────────────────────────────────────────╮
  │    Compiling libc v0.2.174                                                                           │
  │    Compiling shlex v1.3.0                                                                            │
  │    Compiling pkg-config v0.3.32                                                                      │
  │    Compiling vcpkg v0.2.15                                                                           │
  │     Checking bitflags v2.9.1                                                                         │
  │     Checking log v0.4.27                                                                             │
  │    Compiling jobserver v0.1.33                                                                       │
  │    Compiling cc v1.2.27                                                                              │
  │    Compiling libz-sys v1.1.22                                                                        │
  │    Compiling libgit2-sys v0.18.5+1.9.4 (/rust/git2-rs/libgit2-sys)                                   │
  │ warning: libgit2-sys@0.18.5+1.9.4: failed to probe system libgit2:                                   │
  │     Checking git2 v0.21.0 (/rust/git2-rs)                                                            │
  │ error[E0658]: use of unstable library feature `inherent_str_constructors`                            │
  │    --> src/blame.rs:207:25                                                                           │
  │     |                                                                                                │
  │ 207 |             Some(sb) => str::from_utf8(sb).map(|s| Some(s)).map_err(|e| e.into()),             │
  │     |                         ^^^^^^^^^^^^^^                                                         │
  │     |                                                                                                │
  │     = note: see issue #131114 <https://github.com/rust-lang/rust/issues/131114> for more information │
  │                                                                                                      │
  │ For more information about this error, try `rustc --explain E0658`.                                  │
  │ error: could not compile `git2` (lib) due to 1 previous error                                        │
  │                                                                                                      │
  ╰──────────────────────────────────────────────────────────────────────────────────────────────────────╯
                                                                                                          

Compatibility Check #3: Rust 1.91.1
  [OK]     Is compatible 

Compatibility Check #4: Rust 1.88.0
  [OK]     Is compatible 

Compatibility Check #5: Rust 1.87.0
  [OK]     Is compatible 

Result:
   Considered (min … max):   Rust 1.56.1 … Rust 1.96.0 
   Search method:            bisect                    
   MSRV:                     1.87.0                    
   Target:                   x86_64-unknown-linux-gnu  
                                                      
```

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -37,15 +37,7 @@ jobs:
     - name: Install Rust (rustup)
       run: rustup update ${{ matrix.rust }} --no-self-update && rustup default ${{ matrix.rust }}
       shell: bash
-    - name: Cargo test (locked)
-      if: ${{ matrix.build != 'MSRV' }}
-      run: cargo test --locked
-    - name: Remove lock file
-      if: ${{ matrix.build == 'MSRV' }}
-      run: rm Cargo.lock
-    - name: Cargo test (unlocked)
-      if: ${{ matrix.build == 'MSRV' }}
-      run: cargo test
+    - run: cargo test --locked
     - run: cargo test --features https,ssh
     # Single-threaded: libgit2's builtin SHA256 is not thread-safe.
     # See https://github.com/rust-lang/git2-rs/issues/1255.

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -9,7 +9,7 @@ jobs:
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        build: [stable, beta, nightly, macos, windows]
+        build: [stable, beta, nightly, macos, windows, MSRV]
         include:
           - build: stable
             os: ubuntu-latest
@@ -20,6 +20,10 @@ jobs:
           - build: nightly
             os: ubuntu-latest
             rust: nightly
+          # 1.87 is the minimum supported per Cargo.toml
+          - build: MSRV
+            os: ubuntu-latest
+            rust: "1.87"
           - build: macos
             os: macos-latest
             rust: stable
@@ -33,7 +37,15 @@ jobs:
     - name: Install Rust (rustup)
       run: rustup update ${{ matrix.rust }} --no-self-update && rustup default ${{ matrix.rust }}
       shell: bash
-    - run: cargo test --locked
+    - name: Cargo test (locked)
+      if: ${{ matrix.build != 'MSRV' }}
+      run: cargo test --locked
+    - name: Remove lock file
+      if: ${{ matrix.build == 'MSRV' }}
+      run: rm Cargo.lock
+    - name: Cargo test (unlocked)
+      if: ${{ matrix.build == 'MSRV' }}
+      run: cargo test
     - run: cargo test --features https,ssh
     # Single-threaded: libgit2's builtin SHA256 is not thread-safe.
     # See https://github.com/rust-lang/git2-rs/issues/1255.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -14,6 +14,7 @@ repositories.
 """
 categories = ["api-bindings"]
 edition = "2021"
+rust-version = "1.87"
 
 [dependencies]
 url = { version = "2.5.4", optional = true }

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -29,7 +29,7 @@ openssl-probe = { version = "0.1", optional = true }
 
 [dev-dependencies]
 clap = { version = "4.4.13", features = ["derive"] }
-time = { version = "0.3.47", features = ["formatting"] }
+time = { version = "0.3.45", features = ["formatting"] }
 tempfile = "3.1.0"
 url = "2.5.4"
 

--- a/git2-curl/Cargo.toml
+++ b/git2-curl/Cargo.toml
@@ -11,6 +11,7 @@ Backend for an HTTP transport in libgit2 powered by libcurl.
 Intended to be used with the git2 crate.
 """
 edition = "2021"
+rust-version = "1.87"
 
 [dependencies]
 curl = "0.4.33"

--- a/libgit2-sys/Cargo.toml
+++ b/libgit2-sys/Cargo.toml
@@ -15,6 +15,7 @@ exclude = [
     "libgit2/tests/*",
 ]
 edition = "2021"
+rust-version = "1.87"
 
 [lib]
 name = "libgit2_sys"

